substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/code/CodeInfoEncoder.java

```diff
@@ -393,6 +577,286 @@ public boolean verifyFrameInfo(CodeInfo info) {
+                buf.putU2(array.length);
+                for (char val : array) {
+                    encodeValue(buf, val, componentType);
+                }
+            } else if (componentType == int.class) {
+                int[] array = (int[]) value;
+                buf.putU2(array.length);
+                for (int val : array) {
+                    encodeValue(buf, val, componentType);
+                }
+            } else if (componentType == long.class) {
+                long[] array = (long[]) value;
+                buf.putU2(array.length);
+                for (long val : array) {
+                    encodeValue(buf, val, componentType);
+                }
+            } else if (componentType == float.class) {
+                float[] array = (float[]) value;
+                buf.putU2(array.length);
+                for (float val : array) {
+                    encodeValue(buf, val, componentType);
+                }
+            } else if (componentType == double.class) {
+                double[] array = (double[]) value;
+                buf.putU2(array.length);
+                for (double val : array) {
+                    encodeValue(buf, val, componentType);
+                }
+            }
+        }
+
+        byte tag(Class<?> type) {
+            if (type.isAnnotation()) {
+                return '@';
+            } else if (type.isEnum()) {
+                return 'e';
+            } else if (type.isArray()) {
+                return '[';
+            } else if (type == Class.class) {
+                return 'c';
+            } else if (type == String.class) {
+                return 's';
+            } else if (type.isPrimitive()) {
+                return (byte) Wrapper.forPrimitiveType(type).basicTypeChar();
+            } else if (Wrapper.isWrapperType(type)) {
+                return (byte) Wrapper.forWrapperType(type).basicTypeChar();
+            } else {
+                throw shouldNotReachHere();
+            }
+        }
+
+        private Annotation[] filterAnnotations(Annotation[] annotations) {
+            List<Annotation> filteredAnnotations = new ArrayList<>();
+            for (Annotation annotation : annotations) {
+                Class<? extends Annotation> annotationClass = annotation.annotationType();
+                if (supportedValue(annotationClass, annotation, null)) {
+                    filteredAnnotations.add(annotation);
+                }
+            }
+            return filteredAnnotations.toArray(new Annotation[0]);
+        }
+
+        private void registerStrings(Annotation[] annotations) {
+            for (Annotation annotation : annotations) {
+                List<String> stringValues = new ArrayList<>();
+                if (supportedValue(annotation.annotationType(), annotation, stringValues)) {
+                    for (String stringValue : stringValues) {
+                        encoders.sourceMethodNames.addObject(stringValue);
```

Comment:
Interesting that you add to `encoders.sourceMethodNames` and not `encoders.names`. Is there even a reason to have two string tables? Both tables could contain the same strings, leading to duplication. Are some table indices much smaller when using two separate tables, so that we benefit from fewer bytes in the encoding?

Reply:
The distinction was present already in the code so I don't know the rationale behind it. It seems that `names` is meant for debugging information though so it felt more appropriate to add it to the encoder that is used for code that is actually required for correctness. The naming is perhaps not ideal then, that's true.